spec/controllers/slack_controller_spec.rb

```diff
@@ -0,0 +1,153 @@
+        new_tags = ["test", "example"]
+        xhr :post, :edit, {id: id, channel: new_channel, category_id: new_category_id, filter: 'watch', tags: new_tags}
+        expect(response).to be_success
+        filter = ::DiscourseSlack::Slack.get_filter(id)
+        expect(filter[:channel]).to eq(new_channel)
+        expect(filter[:category_id]).to eq(new_category_id)
+        expect(filter[:filter]).to eq('watch')
+        expect(filter[:tags]).to eq(new_tags)
+        cid = ::PluginStore.get(PLUGIN_NAME, "_category_#{new_category_id}_#{new_channel}")
+        expect(cid).to eq(id)
+        new_tags.each do |t|
+          tid = ::PluginStore.get(PLUGIN_NAME, "_tag_#{t}_#{new_channel}")
+          expect(tid).to eq(id)
+        end
+      end
+
+      it "updates the filter without tags" do
+        tags = ["test", "example"]
+        id = ::DiscourseSlack::Slack.set_filter("#hello", "follow", 1, tags)
+        new_channel = "welcome"
+        new_category_id = "2"
+        xhr :post, :edit, {id: id, channel: new_channel, category_id: new_category_id, filter: 'watch', tags: []}
+        expect(response).to be_success
+        filter = ::DiscourseSlack::Slack.get_filter(id)
+        expect(filter[:channel]).to eq(new_channel)
+        expect(filter[:category_id]).to eq(new_category_id)
+        expect(filter[:filter]).to eq('watch')
+        expect(filter[:tags]).to eq([])
+        cid = ::PluginStore.get(PLUGIN_NAME, "_category_#{new_category_id}_#{new_channel}")
+        expect(cid).to eq(id)
+        tags.each do |t|
+          tid = ::PluginStore.get(PLUGIN_NAME, "_tag_#{t}_#{new_channel}")
+          expect(tid).to eq(nil)
+        end
+      end
+
+    end
+
+    context 'command' do
+
+      before do
+        SiteSetting.slack_incoming_webhook_token = "SECRET TOKEN"
+      end
+
+      it 'should create filter to follow a category on slack' do
+        category = Fabricate(:category)
+        expect {
+          xhr :post, :command, {text: "follow #{category.slug}", channel_name: "welcome", token: "SECRET TOKEN"}
+          expect(response).to be_success
+        }.to change(PluginStoreRow, :count).by(2)
+      end
+
+      it 'should create filter to watch a tag on slack' do
+        tag = Fabricate(:tag)
+        expect {
+          xhr :post, :command, {text: "watch tag:#{tag.name}", channel_name: "welcome", token: "SECRET TOKEN"}
+          expect(response).to be_success
+        }.to change(PluginStoreRow, :count).by(2)
+      end
+
+      it 'should remove filter to mute a category on slack' do
+        expect {
+          xhr :post, :command, {text: "mute all", channel_name: "general", token: "SECRET TOKEN"}
+          expect(response).to be_success
+        }.to change(PluginStoreRow, :count).by(1)
+      end
+
+      it 'should remove filter to mute a tag on slack' do
+        tag = Fabricate(:tag)
+        xhr :post, :command, {text: "watch tag:#{tag.name}", channel_name: "welcome", token: "SECRET TOKEN"}
+        expect {
```

Comment:
Style: Generally we leave a blank line before and after code blocks that span multiple lines. Otherwise, there is no clear separation between our code blocks making it hard to read.